NP- 327 emscripten communication

jellespijker opened the pull request:
# Description

Add an extra communication option when using Emscripten and compiling for wasm. This is based on the cmdline communication layer and uses a callback function to be executed on the JS side, providing slice information and possibly in the future preview information.

## Type of change

<!-- Please delete options that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [X] Locally UT, for Arcus, Commandline and Emscripten.
- [X] In the CuraEngineJS

**Test Configuration**:
* Operating System: wasm

# Checklist:

- [X] My code follows the style guidelines of this project as described in [UltiMaker Meta](https://github.com/Ultimaker/Meta)
- [X] I have read the [Contribution guide](https://github.com/Ultimaker/Cura/blob/main/contributing.md)
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have uploaded any files required to test this change

jellespijker committed:
Use smart pointer for communication

Contribute to NP-327

jellespijker committed:
Introduce EmscriptenCommunication class

Contribute to NP-327

jellespijker committed:
Enable Emscripten support in communication module

Refactored conditional compilation to enable Emscripten-specific behavior in the communication module. Updated CMakeLists.txt to handle Emscripten build settings and refactored EmscriptenCommunication to properly manage progress handling and command-line arguments.

Contribute to NP-327

jellespijker committed:
Applied clang-format.

jellespijker committed:
Enable slice info callback for Emscripten communication

This commit adds support for handling slice info callbacks through Emscripten, alongside progress callbacks. Additionally, functions to create slice info messages using rapidjson were implemented to prepare JSON data for slice information.

Contribute to NP-327

jellespijker committed:
Switch to smart pointers

Replaced raw pointers with `std::shared_ptr` for better memory management. Standardized variable names to follow consistent naming conventions, improving code readability and maintainability.

This was needed because the reference to Slice was otherwise out-of-scope when calling `Emscripten::sliceNext()`

Contribute to NP-327

jellespijker committed:
Applied clang-format.

jellespijker committed:
Switch to shared_ptr for better memory management

Replaced unique_ptr with shared_ptr in various components to facilitate shared ownership of resources and prevent manual memory management. This change affects communication objects, `Slice` instances, and related test cases to ensure smooth integration.

Contribute to NP-327

jellespijker committed:
Merge remote-tracking branch 'origin/NP-327_emscripten_communication' into NP-327_emscripten_communication

jellespijker committed:
Merge branch 'main' into NP-327_emscripten_communication

wawanbreton commented on src/communication/CommandLine.cpp:
```diff
@@ -204,14 +193,12 @@ void CommandLine::sliceNext()
-                else if (argument.find("--progress") == 0)
+                else if (argument.starts_with("--progress_cb") || argument.starts_with("--slice_info_cb"))
                 {
-                    // Store progress handler name
                     argument_index++;
                     argument = arguments_[argument_index];
-                    progressHandler = argument;
                 }
-#endif
+#endif // __EMSCRIPTEN__
```
Shouldn't this part be moved to the dedicated class ?

jellespijker replied:
_This is me, trying very hard not to rewrite our custom command line parsing logic_. I normally agree with you. But solving this properly would be more work, this was the quickest fix.

I am tempted to use cxxopts https://github.com/jarro2783/cxxopts an open a community PR one of these days.

jellespijker replied:
Let me know how strong you feel about letting this through

wawanbreton replied:
Well it just feels sad that you took a lot of effort to split the Emscripten part out of CommandLine, and yet you still have this dirty part remaining there :smiley:
You could solve this easily by having a virtual method to parse extra arguments, that would return false by default and would parse those in the `EmscriptenCommunication` overridden definition.

jellespijker replied:
Yes, it feels dirty, Let's see if I can manage to find some time to make the changes for a cleaner solution

wawanbreton commented on src/communication/EmscriptenCommunication.cpp:
```diff
@@ -0,0 +1,111 @@
+    }
+    doc.AddMember("material_estimates", material_estimates_json, allocator);
+
+    // Set CureEngine information
```
```suggestion
    // Set CuraEngine information
```

wawanbreton commented on stress_benchmark/stress_benchmark.cpp:
```diff
@@ -195,7 +195,7 @@ void createAndWriteJson(const std::filesystem::path& out_file, double stress_lev
     rapidjson::Document doc;
     doc.SetArray();
-    rapidjson::Document::AllocatorType& allocator = doc.GetAllocator();
+        rapidjson::Document::AllocatorType& allocator = doc.GetAllocator();
```
This added indentation looks wrong :thinking:

wawanbreton reviewed (commented):
Oooooooooh
Nice improvement, some minor remarks but apart from that I do like it very much :+1:

jellespijker commented on src/Application.cpp:
```diff
@@ -214,8 +215,11 @@ void Application::slice()
+    communication_ = std::make_shared<EmscriptenCommunication>(arguments);
+#else
+    communication_ = std::make_shared<CommandLine>(arguments);
+#endif
```
I don't see a reason why, there is after all a tight coupling with https://github.com/Ultimaker/CuraEngineJS/pull/5 
Maybe a third-party would want to run it with just a command line interface as wasm binary, be we won't.

wawanbreton replied:
It's just to make sure that you won't later get into a situation where you would actually want to do it. But if it really does not make sense, then don't allow it.

jellespijker committed:
Document EmscriptenCommunication class.

Added detailed documentation comments for the EmscriptenCommunication class, including descriptions for member variables and methods. This improves code readability and ease of maintenance.

Contribute to NP-327

jellespijker committed:
Merge remote-tracking branch 'origin/NP-327_emscripten_communication' into NP-327_emscripten_communication

jellespijker committed:
Apply review suggestions

Contribute to NP-327

Commit:
build: add bigint flag to cmake for emscripten

Commit:
refactor(FffGcodeWriter): fix extruder order calculation bug

The return type of getTotalExtraLayer function on Raft is casted to the value_type to prevent underflow. This previously resulted in method xl to generate invalid gcode.

0x5844 commented:
https://github.com/Ultimaker/CuraEngine/pull/2131/commits/6697713ce16a799646e093fb10866eb70684f2d6 is commited to handle [NP-343](https://ultimaker.atlassian.net/browse/NP-343) "Slicing for a method XL produces unexpected results" where the generated gcode was empty. The code change is under #2134

saumyaj3 committed:
Add GCode prefix handling in Emscripten communication

NP-351

saumyaj3 committed:
Applied clang-format.

saumyaj3 committed:
fixing typo in .h file

NP-351

jellespijker committed:
Np 351 adding gcode header to emscription (#2136)

jellespijker committed:
Merge branch 'main' into NP-327_emscripten_communication

jellespijker committed:
Apply review suggestions

Contribute to NP-327

wawanbreton commented:
There has been many commits on this PR recently, is it ready for review again ? Unit tests are failing though.

jellespijker committed:
Fix UT

Contribute to NP-327

jellespijker committed:
Merge remote-tracking branch 'origin/NP-343-slicing-for-a-method-xl-produces-unexpected-results' into NP-327_emscripten_communication

jellespijker commented:
@wawanbreton This now also integrate https://github.com/Ultimaker/CuraEngine/pull/2134 since we're building our Working package of this branch

jellespijker committed:
Communication isn't sequential

Contribute to NP-327

jellespijker committed:
Applied clang-format.

wawanbreton commented on include/communication/EmscriptenCommunication.h:
```diff
@@ -0,0 +1,65 @@
+{
+private:
+    std::string progress_handler_; ///< Handler for progress messages.
+    std::string gcode_prefix_handler_; /// <Handler for getting gcode_prefix.
```
The "prefix" naming is a bit confusing, if I'm correct we mostly use "header" for this, in the front-end and in the engine (although it seems inconsistent in the engine yet). Would you mind renaming the methods/variables to "header" ?

wawanbreton reviewed (commented):
Just one nitpicky remark about naming, which got me confused. Otherwise looks good.

jellespijker committed:
Refactor GCode handler variable and command line flag names

Renamed `gcode_prefix_handler_` to `gcode_header_handler_` for clarity. Updated corresponding usage in `EmscriptenCommunication.h`, `CommandLine.cpp`, and `EmscriptenCommunication.cpp`.

Contribute to NP-327

saumyaj3 committed:
Fix GCode prefix formatting in EmscriptenCommunication

Ensure the prefix argument is correctly formatted as a string in JavaScript function calls. This change prevents potential issues with GCode prefix handling in the communication module as it starts with ';'

NP-327

saumyaj3 committed:
Fix GCode prefix handler name in EmscriptenCommunication

NP-327

Commit:
refactor: remove unnecessary type cast

Commit:
Merge branch 'NP-327_emscripten_communication' of github.com:Ultimaker/CuraEngine into NP-327_emscripten_communication

0x5844 commented:
@wawanbreton @jellespijker I removed the type cast; I gave it additional tests and it seems that the `WASM_BIGINT` flag is sufficient to handle slicing issues with method xl.

0x5844 commented on include/communication/EmscriptenCommunication.h:
```diff
@@ -0,0 +1,65 @@
+     */
+    void sliceNext() override;
+
+    bool isSequential() const
```
```shell
EmscriptenCommunication.h:56:10: warning: 'isSequential' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]
   56 |     bool isSequential() const
      |          ^
CommandLine.h:56:10: note: overridden virtual function is here
   56 |     bool isSequential() const override;
      |          ^
1 warning generated.
```

saumyaj3 committed:
Add base64 encoding for GCode prefix handling.

Implemented a `convertTobase64` function to encode strings to base64, ensuring they can be safely transmitted in JavaScript. Updated `EmscriptenCommunication::sendGCodePrefix` to use this new encoding method and modified relevant tests accordingly to check this functionality.

NP-327

saumyaj3 committed:
Applied clang-format.

saumyaj3 committed:
Rename handler for GCode prefix transmission.

NP-327

jellespijker committed:
Update include/communication/EmscriptenCommunication.h

Co-authored-by: Erwan MATHIEU <erwan.mathieu@ultimaker.com>

saumyaj3 committed:
Refactor to use existing mesh group object

assuming that cloud slicer does not create multiple mesh groups, but one.

NP-351

wawanbreton commented:
What is the status of this PR ? There has been a few significant changes since the last approval (which I would ask you to avoid as much as possible), is it now stable enough for a final review ?

jellespijker commented:
Yes it can be reviewed as final

casperlamboo commented:
> @wawanbreton @jellespijker I removed the type cast; I gave it additional tests and it seems that the WASM_BIGINT flag is sufficient to handle slicing issues with method xl.

With these proposed changes I am still able to reproduce the method xl slicing issues. I think the cast is still needed for wasm.

0x5844 commented:
> > @wawanbreton @jellespijker I removed the type cast; I gave it additional tests and it seems that the WASM_BIGINT flag is sufficient to handle slicing issues with method xl.
> 
> With these proposed changes I am still able to reproduce the method xl slicing issues. I think the cast is still needed for wasm. (FYI @0x5844)

That's strange, if we want to keep that cast we have to adjust it in all it's occurrences (I don't think this is the way to go about it). Currently there are multiple changes to how we compile wasm (flags passed to the compiler); I would much  rather wait for the merge and have a fresh test again. Removing the typecast while enabling the compiler flag seemed to tackle to issue (?) which seem to be no longer the case (this was tested last week).

HellAholic committed:
Np 351 header containing extruder info (#2138)
